src/librustc/mir/repr.rs

```diff
@@ -962,7 +962,10 @@ pub enum CastKind {
 pub enum AggregateKind<'tcx> {
     Vec,
     Tuple,
-    Adt(AdtDef<'tcx>, usize, &'tcx Substs<'tcx>),
+    /// The second field is variant number (discriminant), it's equal to 0
+    /// for struct and union expressions. The fourth field is active field
+    /// number and is present only for union expressions.
+    Adt(AdtDef<'tcx>, usize, &'tcx Substs<'tcx>, Option<usize>),
```

Comment:
Yeah, it's a consequence of unions being univariant `AdtDef`s, the second `usize` is used for indexing into `AdtDef`'s variants all the time, it's more convenient to keep it `0`, like for `struct`s.
(I'm not proud of this, but this is the less invasive change.)